pkg/kv/postgres/store.go

```diff
@@ -0,0 +1,239 @@
+		TableName: defaultTableName,
+	}
+	if tableName, ok := runtimeParams[paramTableName]; ok {
+		p.TableName = tableName
+	}
+	return p
+}
+
+func migrateDatabase(ctx context.Context, conn *pgxpool.Conn, params *Params) error {
+	_, err := conn.Exec(ctx, `CREATE TABLE IF NOT EXISTS `+params.TableName+` (
+    key BYTEA NOT NULL PRIMARY KEY,
+    value BYTEA NOT NULL);`)
+	return err
+}
+
+func (s *Store) Get(ctx context.Context, key []byte) ([]byte, error) {
+	if key == nil {
+		return nil, kv.ErrMissingKey
+	}
+	row := s.Pool.QueryRow(ctx, `SELECT value FROM `+s.Params.TableName+` WHERE key = $1`, key)
+	var val []byte
+	err := row.Scan(&val)
+	if errors.Is(err, pgx.ErrNoRows) {
+		return nil, kv.ErrNotFound
+	}
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return val, nil
+}
+
+func (s *Store) Set(ctx context.Context, key, value []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	_, err := s.Pool.Exec(ctx, `INSERT INTO `+s.Params.TableName+`(key,value) VALUES($1,$2)
+			ON CONFLICT (key) DO UPDATE SET value = $2`, key, value)
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return nil
+}
+
+func (s *Store) SetIf(ctx context.Context, key, value, valuePredicate []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	if value == nil {
+		return kv.ErrMissingValue
+	}
+	var (
+		res pgconn.CommandTag
+		err error
+	)
+	if valuePredicate == nil {
+		// use insert to make sure there was no previous value before
+		res, err = s.Pool.Exec(ctx, `INSERT INTO `+s.Params.TableName+`(key,value) VALUES($1,$2) ON CONFLICT DO NOTHING`, key, value)
+	} else {
+		// update just in case the previous value was same as predicate value
+		res, err = s.Pool.Exec(ctx, `UPDATE `+s.Params.TableName+` SET value=$2 WHERE key=$1 AND value=$3`, key, value, valuePredicate)
+	}
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	if res.RowsAffected() != 1 {
+		return kv.ErrNotFound
+	}
+	return nil
+}
+
+func (s *Store) Delete(ctx context.Context, key []byte) error {
+	if key == nil {
+		return kv.ErrMissingKey
+	}
+	res, err := s.Pool.Exec(ctx, `DELETE FROM `+s.Params.TableName+` WHERE key=$1`, key)
+	if err != nil {
+		return fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	if res.RowsAffected() != 1 {
+		return kv.ErrNotFound
+	}
+	return nil
+}
+
+func (s *Store) Scan(ctx context.Context, start []byte) (kv.Entries, error) {
+	var (
+		rows pgx.Rows
+		err  error
+	)
+	if start == nil {
+		rows, err = s.Pool.Query(ctx, `SELECT key,value FROM `+s.Params.TableName+` ORDER BY key`)
+	} else {
+		rows, err = s.Pool.Query(ctx, `SELECT key,value FROM `+s.Params.TableName+` WHERE key >= $1 ORDER BY key`, start)
+	}
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", kv.ErrOperationFailed, err)
+	}
+	return &Entries{
+		rows: rows,
+	}, nil
+}
+
+func (s *Store) Close() {
+	if s.closed {
```

Comment:
This value anywhere except here, wouldn't it result with errors if you call other methods when the `Store` is closed?

Reply:
Each store holds a connection pool - calling close will close the pool. This code is just protecting our code from crash in case we call to Close twice.
It is helpful in cases that we will want to wrap the Store and call to 'Close' also from the wrapper.

Reply:
Note however that there is still a race.  If this is really important we probably want to protect with an atomic op: the tricky cases of closing twice will involve two goroutines.